### PR TITLE
Add ContextManager based profiling namespace API

### PR DESCRIPTION
**Pull request by jysohn23**
Sample PT/XLA client trace (annotations added):
image

### CODE REVIEW

File: test/test_train_mp_mnist.py [hunk 1]
```diff
@@ -129,14 +129,16 @@ def train_mnist(flags, **kwargs):
...
-      loss.backward()
-      xm.optimizer_step(optimizer)
+      with xp.StepTrace('train_mnist', step_num=step):
+        with xp.Trace('build_graph'):
+          optimizer.zero_grad()
+          output = model(data)
+          loss = loss_fn(output, target)
+          loss.backward()
+        # For profiling purposes we're moving the mark step out of iterator
+        xm.optimizer_step(optimizer, barrier=True)
```
Comment (taylanbil) [inline test/test_train_mp_mnist.py:141]: why barrier here? is it because we don't mark step and the profiler will spit out wrong info otherwise?

File: torch_xla/debug/profiler.py [hunk 2]
```diff
@@ -68,3 +68,50 @@ def trace(service_addr: str,
...
+
+  with xp.Trace('fwd_context'):
+    model(input)
+    xm.mark_step()
+  ```
+  """
+  pass
+
+
+class StepTrace(torch_xla._XLAC.profiler.TraceMe):
```
Comment (taylanbil) [inline torch_xla/debug/profiler.py:97]: Why do we separate this class? Can't we use `Trace` at all times?

File: test/test_train_mp_mnist.py (hunk 1 above)
Comment (taylanbil) [reply test/test_train_mp_mnist.py:141]: I see you removed mark_step from iterator. ok.

This usage feels counter-intuitive to me. Can we handle this in the __enter__, __exit__ methods of the decorator?

I'm not sure if we can make the iterator not mark_step (even if it does, fine, whatever), but we can maybe mark_step at __exit__ ? Otherwise, the more complex steps get, the more bookkeeping the user has to do to mark_step at the right time etc. For example, in dlrm, we have 2 optimizer_step's per model step. WDYT?

**Review by taylanbil — COMMENTED** [review]
`train_test_mp*`'s are supposed to be minimal examples of getting torch_xla to work. This file is becoming a bit complex imho. I'd vote for making a whole new file w/ name indicating the usage of profiler. WDYT?

File: test/test_train_mp_mnist.py (hunk 1 above)
Comment (jysohn23) [reply test/test_train_mp_mnist.py:141]: As discussed offline, I'll see if there are cleaner ways from a UX perspective on the profiling experience for these context manager trace marker APIs in terms of moving around `mark_step` calls.

File: torch_xla/debug/profiler.py (hunk 2 above)
Comment (jysohn23) [reply torch_xla/debug/profiler.py:97]: It's a bit of magic on xprof's part but basically when it comes to performance analysis on xprof, they depend on us to provide explicit step markers as "root". This is one by `_r=1` *kwarg: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/profiler/utils/xplane_schema.cc#L169

Comment (taylanbil) [reply torch_xla/debug/profiler.py:97]: We can have one single `Trace` class, and mark_step by default. We can then use it as:

```python
with xp.Trace('trainstep', step_num=i):
  with xp.Trace('graphbuild', markstep=False):
    ...
```

Or we can have a `StepTrace` class that is equivalent to `Trace` with always `markstep=True`. Imho the first is cleaner. We're forcing users to know the difference b/w marking and not marking the step anyway. wdyt?

File: test/test_train_mp_mnist.py [hunk 3]
```diff
@@ -156,7 +158,10 @@ def test_loop_fn(loader):
     accuracy = xm.mesh_reduce('test_accuracy', accuracy, np.mean)
     return accuracy
 
-  train_device_loader = pl.MpDeviceLoader(train_loader, device)
+  # For profiling we're moving the mark step out of iterator
+  # i.e., batches_per_execution=-1
+  train_device_loader = pl.MpDeviceLoader(
+      train_loader, device, batches_per_execution=-1)
```
Comment (taylanbil) [inline test/test_train_mp_mnist.py:164]: discussed offline but noting here too:

adding kwargs like this `batches_per...` in order to get a clean profile is not user friendly. Let's figure out a way to automatically mark step at the end of the `StepTrace`, and not step anywhere else.

**Comment by jysohn23** [comment]
Per having a single `Trace` and `markstep=True`, that wont work since a user may want to profile many parts of a single step, meaning that if we're always marking steps, we're prematurely evaluating. Also per having two different classes this also:

> It's a bit of magic on xprof's part but basically when it comes to performance analysis on xprof, they depend on us to provide explicit step markers as "root". This is one by _r=1 *kwarg: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/profiler/utils/xplane_schema.cc#L169

File: test/test_train_mp_mnist.py (hunk 3 above)
Comment (jysohn23) [reply test/test_train_mp_mnist.py:164]: Yeah seems a bit trickier than it seems as the step marking is embedded within the loader (which I'm not sure about whether that's the first place for it to be in the first place) + multithreading.

Comment (taylanbil) [reply test/test_train_mp_mnist.py:164]: what's the worry about multithreading?

File: torch_xla/debug/profiler.py (hunk 2 above)
Comment (jysohn23) [reply torch_xla/debug/profiler.py:106]: I personally think that it might be a good idea to keep the `TraceStep` pretty explicit given the large behavior difference between the two when it comes to tracing/marking. If you have strong opinions I can change.

File: test/test_train_mp_mnist.py (hunk 3 above)
Comment (jysohn23) [reply test/test_train_mp_mnist.py]: Updated the API, which now shouldn't require any changes from the user in terms of moving where the step is being marked.

File: torch_xla/debug/profiler.py (hunk 2 above)
Comment (taylanbil) [reply torch_xla/debug/profiler.py:109]: yeah ok let's keep it different.

File: test/test_train_mp_mnist.py (hunk 3 above)
Comment (taylanbil) [reply test/test_train_mp_mnist.py]: will look, thanks.

File: torch_xla/distributed/parallel_loader.py [hunk 4]
```diff
@@ -36,11 +37,16 @@ def __len__(self):
...
   def next(self):
-    if self._mark_step_batch_count == self._batches_yielded:
-      self._batches_yielded = 0
-      xm.mark_step()
-    else:
+    if xp.get_tracer_marked_step():
+      xp.set_tracer_marked_step(False)
       self._batches_yielded += 1
+    else:
+      if self._mark_step_batch_count == self._batches_yielded:
```
Comment (taylanbil) [inline torch_xla/distributed/parallel_loader.py:44]: If profiler is marking step (say, for 100 iters), `self._batches_yielded` might grow too large and then we may never end up marking step afterwards, w/ the `==` check. Let's change to `<=`.

File: test/test_profile_mp_mnist.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+  model = MNIST().to(device)
+  writer = None
+  if xm.is_master_ordinal():
+    writer = test_utils.get_summary_writer(flags.logdir)
+  optimizer = optim.SGD(model.parameters(), lr=lr, momentum=flags.momentum)
+  loss_fn = nn.NLLLoss()
+
+  # Start up client side profiler server.
+  server = xp.start_server(flags.profiler_port)
+  # Testing purpose only: set event for synchronization.
```
Comment (taylanbil) [inline test/test_profile_mp_mnist.py:127]: I can't tell if this is needed. If it is not, can we remove for the sake of keeping only the minimal bits to enable the profiler?

Comment (taylanbil) [reply test/test_profile_mp_mnist.py:127]: I mean the below two lines, not start_server obviously.

File: test/test_profile_mp_mnist.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+  if kwargs.get('worker_started'):
+    kwargs.pop('worker_started').set()
+
+  def train_loop_fn(loader):
+    tracker = xm.RateTracker()
+    model.train()
+    for step, (data, target) in enumerate(loader):
+      with xp.StepTrace('train_mnist', step_num=step):
+        with xp.Trace('build_graph'):
```
Comment (taylanbil) [inline test/test_profile_mp_mnist.py:136]: much better, thank you!

File: test/test_profile_mp_mnist.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+          optimizer.zero_grad()
+          output = model(data)
+          loss = loss_fn(output, target)
+          loss.backward()
+        xm.optimizer_step(optimizer)
+      tracker.add(flags.batch_size)
+      if step % flags.log_steps == 0:
```
Comment (taylanbil) [inline test/test_profile_mp_mnist.py:143]: should this go into the StepTrace context? Otherwise, aren't we giving the training update the  next time it steps?

File: test/test_profile_mp_mnist.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        xm.add_step_closure(
+            _train_update, args=(device, step, loss, tracker, writer))
+
+  def test_loop_fn(loader):
```
Comment (taylanbil) [inline test/test_profile_mp_mnist.py:147]: we could remove the test_loop entirely since it's not profiler related, but I'll leave that up to you.

File: torch_xla/debug/profiler.py [hunk 9]
```diff
@@ -68,3 +80,59 @@ def trace(service_addr: str,
       duration_ms=duration_ms,
       num_tracing_attempts=num_tracing_attempts,
       options=options)
+
+
+class Trace(torch_xla._XLAC.profiler.TraceMe):
```
Comment (taylanbil) [inline torch_xla/debug/profiler.py:85]: perhaps beyond the scope of this pr, but we may think about having an env var, that controls the number of steps to profile for. What happens if I run the test_train script above for too long? Are we relying on the user ctrl-c'ing to stop profiling? WDYT about raising an error in the __enter__ or __exit__ if the metric counter for `MarkStep` is greater than or equal to some value?

File: test/test_profile_mp_mnist.py (hunk 5 above)
Comment (taylanbil) [reply test/test_profile_mp_mnist.py:127]: ok

**Review by taylanbil — APPROVED** [review]

**Review by JackCaoG — APPROVED** [review]
Thanks @jysohn23 !

**Comment by jysohn23** [comment]
Thanks for the reviews!
